### PR TITLE
Refactoring Jenkins build scripts

### PR DESCRIPTION
**Pull request by mapellidario**
### status

This is ready to be merged.

this PR contains the following changes
- [x] introduction of `$RPM_RELEASETAG_HASH`, moving to github the jenkins scripts for building of rpm and docker. not deployment.
- [ ] (postponed) ~~moving `CRABServer/Docker` into some other directory~~

next steps (after merge)

- [ ] fix documentation (see below)
- [ ] right after merge, test this PR with browsing https://cmssdt.cern.ch/dmwm-jenkins/job/github-webhook/configure, scrolling down to "Trigger/call build on other projects/Projects to build" and change `CRABServer_BuildOnRelease` with `CRABServer_BuildOnRelease_20220127`
- [ ] when we feel confident with it, we can
    - [ ] copy the content of `CRABServer_BuildOnRelease_20220127` into `CRABServer_BuildOnRelease` 
    - [ ]  copy the content of `CRABServer_BuildImage_20220127` into `CRABServer_BuildImage`
    - [ ] in `github-webhook`, replace `CRABServer_BuildOnRelease_20220127` with `CRABServer_BuildOnRelease`
- [ ] new PR about moving to GH also the scripts for deployment of `crabserver` and `crabtaskworker`
 
### description

I understand this PR may have been bigger than what we envisioned at first, but I hope i improved clarity and maintainability at the expense of verbosity.

If we do not want to deal with this when merging the `python3` branch into the `master` one, we apply some temporary patches to fix the current problems with the build, then come back to this PR after the merge.

How to test:

1. browse https://cmssdt.cern.ch/dmwm-jenkins/job/github-webhook/configure, scroll down to "Trigger/call build on other projects/Projects to build" and change `CRABServer_BuildOnRelease` with `CRABServer_BuildOnRelease_20220127`
2. rebuild last with https://cmssdt.cern.ch/dmwm-jenkins/job/github-webhook/lastCompletedBuild/rebuild/parameterized 
  - repo: `dmwm/CRABServer`
  - event: `release`
  - action: `created`
  - title: `py3.220105`, not the latest, so that you dont risk ruining new/current images
  - payload: this is an example https://dmapelli.web.cern.ch/public/crab/20220127/crabserver_github_release_payload_example.json, but the only important thing is that it contains ` "target_commitish": "python3"` or ` "target_commitish": "master"`, depending which branch the tag you specified is in

Description of the new building:

- github release -> sends webhook
- webhook starts jenkins job https://cmssdt.cern.ch/dmwm-jenkins/job/github-webhook/
  - this job has many inputs (as env variables), but we can focus only on:
    - `TITLE`: this is the tag name
    - `PAYLOAD`: the whole payload from the webhook, this will be used later
  - this jobs has one output (as env variable):
    - `RELEASE_TAG`: same as `TITLE`
- `github-webhook` triggers https://cmssdt.cern.ch/dmwm-jenkins/job/CRABServer_BuildOnRelease/ (right now, the test job is https://cmssdt.cern.ch/dmwm-jenkins/job/CRABServer_BuildOnRelease_20220127/), which only runs `cicd/build/jenkins_build_rpm.sh`
  - this jobs uses:
    - `RELEASE_TAG`
    - `PAYLOAD` to compute the `BRANCH` to which the selected tag points to
  - this job returns (as env variable):
    - `BRANCH`
    - `RPM_RELEASETAG_HASH`: this is what was called before `VERSION_RPMHASH`. 
    - [ ] I can not find a better way of signalling that this contains both the release tag and the hash added to the rpm. Or maybe we can just use `RPM_TAG`, i do not know, i am out of ideas.
- if `CRABServer_BuildOnRelease` is successfull, it triggers https://cmssdt.cern.ch/dmwm-jenkins/job/CRABServer_BuildImage (right now, the test job is https://cmssdt.cern.ch/dmwm-jenkins/job/CRABServer_BuildImage_20220127), which only runs `cicd/build/jenkins_build_docker.sh`
  - this job uses from env
    - `RPM_RELEASETAG_HASH`
    - `BRANCH`
    - `RELEASE_TAG`

Moreover, the TaskWorker Dockerfile has been changed. ~~First of all, I propose to move the `Docker` directory to `cicd/build/crabtaskworker/docker/Dockerfile`.~~ [edit] this has been postponed

Mainly, the Dockerfile was using two variable for the same purpose: `RELEASE` and `TW_VERSION`. I propose to use two different variables, each one with its own meaning: `RELEASE_TAG` and `RPM_RELEASETAG_HASH`. Both these variables needs to be passed when calling `docker build`, so we may need to change the documentation on how to build the docker image manually. `addGH.sh` uses both variables, `install.sh` uses only `RPM_RELEASETAG_HASH`.

If we like this structure, I propose the following changes

- [ ] move the script in https://cmssdt.cern.ch/dmwm-jenkins/job/CRABServer_Deploy_Server/ to `cicd/deploy/jenkins_deploy_rest.sh`
- [ ] move the script in https://cmssdt.cern.ch/dmwm-jenkins/job/CRABServer_Deploy_TW/ to `cicd/deploy/jenkins_deploy_tw.sh`

### collateral effect

We need to update documentation after we merge this, in particular

- [ ] https://twiki.cern.ch/twiki/bin/view/CMSPublic/CRABK8s
- [ ] https://twiki.cern.ch/twiki/bin/view/CMSPublic/CMSCrabTaskWorker
- [ ] https://twiki.cern.ch/twiki/bin/view/CMSPublic/CRABDocker

### CODE REVIEW

**Comment by belforte** [comment]
thanks Dario. Give me some time to digest though

**Comment by belforte** [comment]
I would like to separate this in two parts:
1. introduce $RPM_RELEASETAG_HASH, which looks fine to me
2. move things to different directories. Here I am a bit unhappy with `cicd/build/crabtaskworker/docker/...` both because it also builds Publisher (and I presume later on image used to run crontabs) and because we have this thinkg that docker files for crabserver are in a different place (CMSKubernetes repository) which I am not sure it is a good idea long term. Would be good to be able to change our K8s image w/o asking Valentin to merge things, or we may also decide to run TW or Publisher (or the monitoring crontabs) in K8s... So I propose to clarify all of this before making changes to directory.

**Review by belforte — CHANGES_REQUESTED** [review]
see https://github.com/dmwm/CRABServer/pull/7016#issuecomment-1034926187

**Comment by mapellidario** [comment]
You are correct, I would avoid changing paths in this PR and leave it for the future. I will update this PR.

And you are also correct, that proposal for new directory organization may seem a bit overkill, but emerged from some thinking about how to re-organize everything. Maybe this is not the place to discuss about it, we can create a new low-priority issue. 

Thanks again Stefano for the review!

**Comment by mapellidario** [comment]
I think this is ready to merge.

I removed the directory change for `Docker` directory. If you want, i can open a new low-prio issue in which we can discuss it.

After merging this PR,we need to update the jenkins job specifications. I will do it when I have green light from you.

**Review by belforte — APPROVED** [review]

**Comment by belforte** [comment]
thanks @mapellidario changes look good to me now. But I am sorry I get lost in the very long description and all the action items. I presume that you did it to help you do the changes, so please do, I am sure you will managed. We will review documentation once balls have stopped rolling.
Sure, can open a low prio issue about rationaliziing CI/CD related scripts and repos, or maybe better a GH project. But we do not have the time to flesh out all the topic in there until we close some of the stuff we have in our hands already.

**Comment by belforte** [comment]
@mapellidario should we merge ? Or do you want to keep it open to use the various checkboxes above ? Maybe you can copy that todo list elsewhere ?

**Comment by belforte** [comment]
thanks Dario, mergin

**Pull request merged**
